overlord/devicestate/handlers_install.go

```diff
@@ -618,3 +631,255 @@ func (m *DeviceManager) doRestartSystemToRunMode(t *state.Task, _ *tomb.Tomb) er
 
 	return nil
 }
+
+func (m *DeviceManager) doFactoryResetRunSystem(t *state.Task, _ *tomb.Tomb) error {
+	st := t.State()
+	st.Lock()
+	defer st.Unlock()
+
+	perfTimings := state.TimingsForTask(t)
+	defer perfTimings.Save(st)
+	// get gadget dir
+	deviceCtx, err := DeviceCtx(st, t, nil)
+	if err != nil {
+		return fmt.Errorf("cannot get device context: %v", err)
+	}
+	gadgetInfo, err := snapstate.GadgetInfo(st, deviceCtx)
+	if err != nil {
+		return fmt.Errorf("cannot get gadget info: %v", err)
+	}
+	gadgetDir := gadgetInfo.MountDir()
+
+	kernelInfo, err := snapstate.KernelInfo(st, deviceCtx)
+	if err != nil {
+		return fmt.Errorf("cannot get kernel info: %v", err)
+	}
+	kernelDir := kernelInfo.MountDir()
+
+	modeEnv, err := maybeReadModeenv()
+	if err != nil {
+		return err
+	}
+	if modeEnv == nil {
+		return fmt.Errorf("missing modeenv, cannot proceed")
+	}
+
+	// bootstrap
+	bopts := install.Options{
+		Mount: true,
+	}
+	encryptionType, err := m.checkEncryption(st, deviceCtx)
+	if err != nil {
+		return err
+	}
+	bopts.EncryptionType = encryptionType
+	useEncryption := (encryptionType != secboot.EncryptionTypeNone)
+	hasMaker := osutil.FileExists(filepath.Join(boot.InstallHostFDESaveDir, "marker"))
+	// TODO verify that the same encryption mechanism is used
+	if hasMaker != useEncryption {
+		prevStatus := "encrypted"
+		if !hasMaker {
+			prevStatus = "unencrypted"
+		}
+		// the original system was (un)encrypted
+		return fmt.Errorf("cannot perform factory reset using different encryption, the original system was %v", prevStatus)
+	}
+
+	model := deviceCtx.Model()
+
+	// make sure that gadget is usable for the set up we want to use it in
+	validationConstraints := gadget.ValidationConstraints{
+		EncryptedData: useEncryption,
+	}
+	var ginfo *gadget.Info
+	timings.Run(perfTimings, "read-info-and-validate", "Read and validate gagdet info", func(timings.Measurer) {
+		ginfo, err = gadget.ReadInfoAndValidate(gadgetDir, model, &validationConstraints)
+	})
+	if err != nil {
+		return fmt.Errorf("cannot use gadget: %v", err)
+	}
+	if err := gadget.ValidateContent(ginfo, gadgetDir, kernelDir); err != nil {
+		return fmt.Errorf("cannot use gadget: %v", err)
+	}
+
+	// run the create partition code
+	logger.Noticef("create and deploy partitions")
+	timings.Run(perfTimings, "factory-reset", "Factory reset", func(tm timings.Measurer) {
+		st.Unlock()
+		defer st.Lock()
+		_, err = installFactoryReset(model, gadgetDir, kernelDir, "", bopts, nil, tm)
+	})
+	if err != nil {
+		return fmt.Errorf("cannot perform factory reset: %v", err)
+	}
+
+	// TODO: splice into a common install/reset helper?
+
+	// keep track of the model we installed
+	err = os.MkdirAll(filepath.Join(boot.InitramfsUbuntuBootDir, "device"), 0755)
+	if err != nil {
+		return fmt.Errorf("cannot store the model: %v", err)
+	}
+	err = writeModel(model, filepath.Join(boot.InitramfsUbuntuBootDir, "device/model"))
+	if err != nil {
+		return fmt.Errorf("cannot store the model: %v", err)
```

Comment:
do we need to remove the `device` directory we just created? If not, it's still better writing a comment about it.

Reply:
no not really, if this fails the whole factory reset fails and only way out is doing a reinstall which wipes this partition